certora/WormholeJoin.spec

```diff
@@ -0,0 +1,770 @@
+    assert(revert12 => lastReverted, "revert12 failed");
+    assert(revert13 => lastReverted, "revert13 failed");
+    assert(revert14 => lastReverted, "revert14 failed");
+    assert(revert15 => lastReverted, "revert15 failed");
+    assert(revert16 => lastReverted, "revert16 failed");
+    assert(revert17 => lastReverted, "revert17 failed");
+    assert(revert18 => lastReverted, "revert18 failed");
+    assert(revert19 => lastReverted, "revert19 failed");
+    assert(revert20 => lastReverted, "revert20 failed");
+    assert(revert21 => lastReverted, "revert21 failed");
+    assert(revert22 => lastReverted, "revert22 failed");
+    assert(revert23 => lastReverted, "revert23 failed");
+    assert(revert24 => lastReverted, "revert24 failed");
+
+    assert(lastReverted => revert1  || revert2  || revert3  ||
+                           revert4  || revert5  || revert6  ||
+                           revert7  || revert8  || revert9  ||
+                           revert10 || revert11 || revert12 ||
+                           revert13 || revert14 || revert15 ||
+                           revert16 || revert17 || revert18 ||
+                           revert19 || revert20 || revert21 ||
+                           revert22 || revert23 || revert24, "Revert rules are not covering all the cases");
+}
+
+rule settle(bytes32 sourceDomain, uint256 batchedDaiToFlush) {
+    env e;
+
+    bool vatLive = vat.live() == 1;
+    int256 debtBefore = debt(sourceDomain);
+
+    uint256 inkBefore;
+    uint256 artBefore;
+    inkBefore, artBefore = vat.urns(ilk(), currentContract);
+
+    uint256 cureBefore = cure();
+
+    uint256 vatDaiJoinBefore = vat.dai(currentContract);
+
+    uint256 amtToPayBack = batchedDaiToFlush <= artBefore ? batchedDaiToFlush : artBefore;
+
+    settle(e, sourceDomain, batchedDaiToFlush);
+
+    int256 debtAfter = debt(sourceDomain);
+
+    uint256 inkAfter;
+    uint256 artAfter;
+    inkAfter, artAfter = vat.urns(ilk(), currentContract);
+
+    uint256 cureAfter = cure();
+
+    uint256 vatDaiJoinAfter = vat.dai(currentContract);
+
+    assert(to_mathint(debtAfter) == to_mathint(debtBefore) - to_mathint(batchedDaiToFlush), "debt has not decreased as expected");
+    assert(vatLive => inkAfter == inkBefore - amtToPayBack, "ink has not decreased as expected");
+    assert(vatLive => artAfter == artBefore - amtToPayBack, "art has not decreased as expected");
+    assert(!vatLive => inkAfter == inkBefore, "ink has not stayed the same as expected");
+    assert(!vatLive => artAfter == artBefore, "art has not stayed the same as expected");
+    assert(vatLive => cureAfter == (artBefore - amtToPayBack) * RAY(), "cure has not been updated as expected");
+    assert(!vatLive => cureAfter == cureBefore, "cure has not stayed the same as expected");
+    assert(vatLive && batchedDaiToFlush > artBefore => vatDaiJoinAfter == vatDaiJoinBefore + (batchedDaiToFlush - artBefore) * RAY(), "join vat dai has not increased as expected 1");
```

Comment:
Missing coverage for `batchedDaiToFlush <= artBefore`. It looks like we could cover all cases by changing to:

```
assert(vatLive => vatDaiJoinAfter == vatDaiJoinBefore + (batchedDaiToFlush - amtToPayBack) * RAY(), "join vat dai has not increased as expected 1");
```

(I tried that out and it seems to work.)